packages/cache/__tests__/hydration.js

```diff
@@ -35,3 +35,145 @@ test('rehydrated styles to head can be flushed', () => {
   cache.sheet.flush()
   expect(document.documentElement).toMatchSnapshot()
 })
+
+test('flushing rehydrated styles in the head only affect styles matching the cache key', () => {
+  const css = `color:hotpink;`
+  const hash = hashString(css)
+  safeQuerySelector('head').innerHTML = [
+    '<style data-emotion="emo 1lrxbo5">.emo-1lrxbo5{color:hotpink;}</style>',
+    '<style data-emotion="css qweqwee">.css-qweqwee{color:red;}</style>'
+  ].join('')
+
+  // this moves emotion style tags at initialization time
+  jest.resetModules()
+  require('@emotion/react')
+
+  const cache = createCache({ key: 'emo' })
+  expect(document.documentElement).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="css qweqwee"
+          data-s=""
+        >
+          .css-qweqwee{color:red;}
+        </style>
+        <style
+          data-emotion="emo 1lrxbo5"
+          data-s=""
+        >
+          .emo-1lrxbo5{color:hotpink;}
+        </style>
+      </head>
+      <body />
+    </html>
+  `)
+
+  cache.sheet.flush()
+  expect(document.documentElement).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="css qweqwee"
+          data-s=""
+        >
+          .css-qweqwee{color:red;}
+        </style>
+      </head>
+      <body />
+    </html>
+  `)
+})
+
+test('Should only hydrate style elements matching the cache key', () => {
+  let css = `color:hotpink;`
+  let hash = hashString(css)
+
+  safeQuerySelector(
+    'body'
+  ).innerHTML = `<style data-emotion="emo ${hash}">.css-${hash}{${css}}</style>`
+
+  const cache = createCache({ key: 'css' })
+
+  expect(cache.inserted).toEqual({})
+  expect(document.documentElement).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="emo 1lrxbo5"
+          data-s=""
+        >
+          .css-1lrxbo5{color:hotpink;}
+        </style>
+      </head>
+      <body />
+    </html>
+  `)
+
+  const cache2 = createCache({ key: 'emo' })
+
+  expect(cache2.inserted).toEqual({ [hash]: true })
+  expect(document.documentElement).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="emo 1lrxbo5"
+          data-s=""
+        >
+          .css-1lrxbo5{color:hotpink;}
+        </style>
+      </head>
+      <body />
+    </html>
+  `)
+})
+
+describe('In production mode', () => {
+  beforeEach(() => {
+    process.env.NODE_ENV = 'production'
+  })
+
+  afterEach(() => {
+    process.env.NODE_ENV = 'test'
+  })
+
+  test('Existing head styles from Emotion 10 should not be moved', () => {
+    safeQuerySelector('head').innerHTML = [
+      '<style data-emotion="css-global"></style>',
+      '<style data-emotion="css"></style>',
```

Comment:
maybe worth adding a comment here that those are specific to Emotion 10 because they do not have `data-s` attribute?